### ISSUE TITLE
`Not Nothing` is `Something`

### ISSUE THREAD

**Issue by iRon7** [PowerShell/PowerShell#24051]
### Summary of the new feature / enhancement

In PowerShell `$Null` and `Nothing` (`Void`, empty array, etc.) are essential the same (see: [Enumerable null](https://learn.microsoft.com/powershell/scripting/learn/deep-dives/everything-about-null#enumerable-null)

To check whether a value is holding something, the general advice is to check whether the value is not `$Null` (nothing), as in [Checking for `$null`](https://learn.microsoft.com/powershell/scripting/learn/deep-dives/everything-about-null#checking-for-null), e.g.:
```PowerShell
$Files = Get-ChildItem
if ($Null -ne $Files) {
    # Do something with $Files
}
```

Meaning, aside from counter-intuitive [Yoda condition](https://en.wikipedia.org/wiki/Yoda_conditions), it contains a [double_negative](https://en.wikipedia.org/wiki/Double_negative) where (`$Null` translates to **nothing** and `-ne` translates to **is not**):

## if `$Files` is not nothing ...

Isn't it better to use a *common positive* condition:

## if `$Files` is something (an `object`) ...

Which translates in PowerShell to:

```PowerShell
if ($Files -is [Object]) { ...
```

And to be consistent:

```PowerShell
if ($Files -isnot [Object]) { ...
```

Which contains a single negative as in `$Null -eq $Files`.

This is more a document issue but it is all based on the assumption that:

### `$Value -is [Object]` ⇔ `$Null -ne $Value`

But I am not 100% sure about that...

```PowerShell
Should -ActualValue ('abc' -is [object]) -be ($Null -ne 'abc')
Should -ActualValue ('abc' -isnot [object]) -be ($Null -eq 'abc')
Should -ActualValue (123 -is [object]) -be ($Null -ne 123)
Should -ActualValue (123 -isnot [object]) -be ($Null -eq 123)
Should -ActualValue (@() -is [object]) -be ($Null -ne @())
Should -ActualValue (@() -isnot [object]) -be ($Null -eq @())
Should -ActualValue ($Null -is [object]) -be ($Null -ne $Null)
Should -ActualValue ($Null -isnot [object]) -be ($Null -eq $Null)
```

One stranger in our midst is of cause the [`automationNull`](https://learn.microsoft.com/en-us/dotnet/api/system.management.automation.internal.automationnull) where these two conditions both falsify (return nothing):

```PowerShell
&{} -is [object]
&{} -isnot [object]
```

which you can't actually be compared with the `$Null` comparison:

```PowerShell
$Null -ne &{}
ParserError:
Line |
   1 |  $Null -ne &{}
     |           ~
     | You must provide a value expression following the '-ne' operator. 
```

which requires brackets, and therefore would be equal to:

```PowerShell
Should -ActualValue ((&{}) -is [object]) -be ($Null -ne (&{}))
Should -ActualValue ((&{}) -isnot [object]) -be ($Null -eq (&{}))
```

**Comment by rhubarb-geek-nz** [comment]
> Isn't it better to use a common positive condition

How about

```
$Files = Get-ChildItem
if ($Files) {
    # Do something with $Files
}
```

That will count as true if any files were returned. Note that PowerShell will unwrap a list of one FileInfo into a single FileInfo object.

**Comment by jhoneill** [comment]
> > Isn't it better to use a common positive condition
> 
> How about
> 
> ```
> $Files = Get-ChildItem
> if ($Files) {
>     # Do something with $Files
> }
> ```
> 
> That will count as true if any files were returned. Note that PowerShell will unwrap a list of one FileInfo into a single FileInfo object.

This a standard pattern and where works were zero, empty string etc. which convert to Boolean false aren't values we want to process.    e.g. with `param ($retries) `  if zero is a valid setting,  `if ($retries)  {Set-Retries $retries}` will never set it to zero.


> In PowerShell `$Null` and `Nothing` (`Void`, empty array, etc.) are essential the same
TBH, that's not quite true;  "Never set", "Returned no objects", "Empty" all convert to Boolean false. 
But neither `'' -eq $null` nor `$null -eq '' ` return true  (empty string isn't null, and null doesn't convert) .  So there are places where we need things like `[string]::isnullOrEmpty ` but `if (-not $)`  will do the same thing if `$s` is null or holds an empty string.

**Comment by iRon7** [comment]
@rhubarb-geek-nz,

> How about

Correct, the "`$files`" example wasn't well chosen but as **jhoneil** points out, it is about a general comparison to confirm that a common `$Value` (and not just `$files`) has been set:

```PowerShell
$Value = Try { Get-Random 3 } Catch { Throw $_ } # Where `Get-Random` could be any function
if ($Value -is [object]) {                       # Which we only want to process if it contains something (including 0, "", etc)
```

@jhoneill

> TBH, that's not quite true; "Never set", "Returned no objects", "Empty" all convert to Boolean false.

Right, my statement is probably a little bluntly but TBH your statement is also not full detailed.

I think it is correct to say that they all "[*evaluate* to `$false`](https://learn.microsoft.com/powershell/module/microsoft.powershell.core/about/about_booleans#converting-from-scalar-types)" or when assigned to a variable some translate to `$Null`:

``` PowerShell
Remove-Variable NeverSet -ErrorAction SilentlyContinue
if ($Null -eq $NeverSet) { '$NeverSet is $Null' }
$NoObjects = &{ <#Returned no objects#> }
if ($Null -eq $NoObjects) { '$NoObjects is $Null' }
```

Anyways (that is not the issue), the issue/question is:
1) Can I use `$Value -is [Object]` for every `$Null -ne $Value` statement?
2) If yes, wouldn't it be better to use `$Value -is [Object]` instead of the *double-negative Yoda condition*?

**Comment by jhoneill** [comment]
> Anyways (that is not the issue), the issue/question is:
> 
> 1. Can I use `$Value -is [Object]` for every `$Null -ne $Value` statement?
> 2. If yes, wouldn't it be better to use `$Value -is [Object]` instead of the _double-negative Yoda condition_?

`$Value -is [Object]`  will fail in the case that $Value is an empty array,  there are cases where we do 
```
$stuff = @() 
$stuff += Get-Things @params1
$stuff +=Get-Things @params2
```
(In this case `$null -eq $stuff` gives $false, and `$stuff -eq $null` gives an empty array which evaluates to Boolean false)
We can test `if ($stuff)`,  or `if ($stuff.count)`, or `if ($stuff.count -gt 0 )`  Since PowerShell returns count =1 for scalars (even zero, false or empty string) you could use .`count -gt 0` as a universal test. 
    

We use double negatives or "is not what we do not want"  quite commonly 
  `$x -ne ''`  ;  `$x.count -ne 0`  etc and usually we can (and do) write `if ($value)` 

Given that it  `if ($x)`  , `if ($x -ne $null)` or the array-safe `if ($null -ne $x)`  are well understood I would avoid creating a new pattern in code to be shared with others, because it requires people to know that things that aren't object are descended from object and will make them ask why you are using a different test 

I would also say that where we need to act on the thing being tested to raise an error the code will often be clearer if we dodge the double negative. 

```
if ($x)  {
   do a lot of stuff that takes a whole page
}
else {report error}
```
Means when we read the else we have forgotten what the if was 

```
if (-not $x)  {report error}
else {
   do a lot of stuff that takes a whole page
}
```
Is a better pattern.   

`If ($x -eq $null)` / `If ($null -eq $x)`  / if ($x -isnot

**Comment by mklement0** [comment]
* I personally think that the double negative isn't problematic, and from the standpoint of conceptual clarity `$null -ne $someValue` strikes me as preferable to `$someValue -is [object]` - but the two should be equivalent.

* The "Yoda syntax" is somewhat problematic, but not specific to this case: it is an unavoidable consequence of how _comparison operators_ work in PowerShell.

For extending the logic to "is this not $null, the empty string, or an empty collection?", the simplest (though not clearest) formulation is probably:

```powershell
if ($someValue.Length) { 'is a nonempty string, a non-string scalar, or a non-empty collection' }
```

This relies on the intrinsic `.Length` property rather than `.Count`, which are equivalent. While `.Count` may be conceptually preferable, using `.Length` ensures that the test also works with _strings_.

(I'm using "collection" loosely to refer to any instance that PowerShell auto-enumerates in the pipeline; for `IEnumerable` instances that aren't also `IList`, e.g. in LINQ expressions such as `([Linq.Enumerable]::Take[int]([int[]] (0, 1, 2), 2)).Length`, `.Length` triggers enumeration and therefore member-access enumeration, so it is applied to each enumerated element; while inefficient, the test should still work as intended).

**Comment by SeeminglyScience** [comment]
I'm going to convert this to a discussion since as far as I can tell there's nothing actionable here. Also might be a good question to pose to the [PowerShell discord](https://aka.ms/psdiscord).

My personal opinion is even if `-is [object]` is consistent (which I don't know the answer to) it's significantly less readable. Partially because I then have to ask if they're using it as a null check or if they have a large misconception about the type system, and partially because when framing it as a reverse null check it forces flipped contexts a bit (e.g. "oh it's used to check for null. But oh right opposite, so `is not null`". Just extra cognitive work imo)
